compiler/rustc_pattern_analysis/src/rustc.rs

```diff
@@ -913,7 +913,8 @@ impl<'p, 'tcx: 'p> PatCx for RustcPatCtxt<'p, 'tcx> {
 
     fn complexity_exceeded(&self) -> Result<(), Self::Error> {
         let span = self.whole_match_span.unwrap_or(self.scrut_span);
-        Err(self.tcx.dcx().span_err(span, "reached pattern complexity limit"))
+        self.tcx.dcx().span_warn(span, "`match` is too complex");
```

Comment:
Could you make this into a proper translatable diagnostic, with a more helpful message?

I'm thinking:
```
warning: this pattern-match expression is taking too long to analyze
help: break it up into smaller `match`es and/or replace some patterns with guards
note: match checking can take exponential time when many different fields of structs/tuples/arrays are involved
```

Reply:
Done as well. :)